# Laboratorio 7

In [7]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import altair as alt

from sklearn import datasets, linear_model
from sklearn.metrics import mean_squared_error, r2_score

alt.themes.enable('opaque')
%matplotlib inline

In [ ]:
En este laboratorio utilizaremos los mismos datos de diabetes vistos en la clase

In [3]:
diabetes_X, diabetes_y = datasets.load_diabetes(return_X_y=True, as_frame=True)
diabetes = pd.concat([diabetes_X, diabetes_y], axis=1)
diabetes.head()

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  target  
0 -0.002592  0.019908 -0.017646   151.0  
1 -0.039493 -0.068330 -0.092204    75.0  
2 -0.002592  0.002864 -0.025930   141.0  
3  0.034309  0.022692 -0.009362   206.0  
4 -0.002592 -0.031991 -0.046641   135.0

## Pregunta 1

(1 pto)

* ¿Por qué la columna de sexo tiene esos valores?
* ¿Cuál es la columna a predecir?
* ¿Crees que es necesario escalar o transformar los datos antes de comenzar el modelamiento?

In [28]:
diabetes.groupby("sex").agg({"sex":"sum"}), -10.490785/-0.044642, 10.490785/0.050680  

(                 sex
 sex                 
 -0.044642 -10.490785
  0.050680  10.490785,
 234.9980959634425,
 207.00049329123914)

__Respuesta:__

*  Porque si bien la variable binaria original sería 0 y 1, de manera arbitraria para un sexo y para el otro, las variables de la base de datos han sido centradas y escaladas por su desviación estandar multiplicada por la cantidad de muestras, para que así tengan una norma unitaria. En particular, los valores corresponden al valor que divide al número de datos de cada sexo, con tal de que la suma de ellos, sea igual para cada categoría: Inicialmente, la suma de un groupby("sex") al dataframe entrega el siguiente valor: |suma|=10.490785, y luego, al divir este por los valores presentes, obtiene la cantidad de datos perteneciente a la categoría: -10.490785/-0.044642= 234.9980959634425, y 10.490785/0.050680=207.00049329123914.
* Target, puesto que es la medida cuantitativa de la progresión de la enfermedad según el paciente.
* Sí, puesto que podrían mostrar datos que no sean fieles a la realidad dependiendo del contexto, es decir, ayuda a la posterior interpretación de ellos; o bien pueden simplemente ayudar a encontrar una predicción que se adecúe a más escenarios que en el caso no transformado y a evitar inestabilidad numérica, pudiendo ser una transformación normalizar el dataset. Esto, además de que el algoritmo puede funcionar mejor, de que se pueden valorizar adecuadamente los datos según su naturaleza y de que se puede otorgar más valor al modelo en el caso de que se eliminen distintas escalas que hayan estado aplicadas en los datos de entrada.

## Pregunta 2

(1 pto)

Realiza dos regresiones lineales con todas las _features_, el primer caso incluyendo intercepto y el segundo sin intercepto. Luego obtén la predicción para así calcular el error cuadrático medio y coeficiente de determinación de cada uno de ellos.

In [27]:
mean_squared_error?

Signature:
mean_squared_error(
    y_true,
    y_pred,
    *,
    sample_weight=None,
    multioutput='uniform_average',
    squared=True,
)
Docstring:
Mean squared error regression loss

Read more in the :ref:`User Guide <mean_squared_error>`.

Parameters
----------
y_true : array-like of shape (n_samples,) or (n_samples, n_outputs)
    Ground truth (correct) target values.

y_pred : array-like of shape (n_samples,) or (n_samples, n_outputs)
    Estimated target values.

sample_weight : array-like of shape (n_samples,), optional
    Sample weights.

multioutput : string in ['raw_values', 'uniform_average']
    or array-like of shape (n_outputs)
    Defines aggregating of multiple output values.
    Array-like value defines weights used to average errors.

    'raw_values' :
        Returns a full set of errors in case of multioutput input.

    'uniform_average' :
        Errors of all outputs are averaged with uniform weight.

squared : boolean value, optional (default = True)
    If

In [29]:
r2_score?

Signature:
r2_score(
    y_true,
    y_pred,
    *,
    sample_weight=None,
    multioutput='uniform_average',
)
Docstring:
R^2 (coefficient of determination) regression score function.

Best possible score is 1.0 and it can be negative (because the
model can be arbitrarily worse). A constant model that always
predicts the expected value of y, disregarding the input features,
would get a R^2 score of 0.0.

Read more in the :ref:`User Guide <r2_score>`.

Parameters
----------
y_true : array-like of shape (n_samples,) or (n_samples, n_outputs)
    Ground truth (correct) target values.

y_pred : array-like of shape (n_samples,) or (n_samples, n_outputs)
    Estimated target values.

sample_weight : array-like of shape (n_samples,), optional
    Sample weights.

multioutput : string in ['raw_values', 'uniform_average', 'variance_weighted'] or None or array-like of shape (n_outputs)

    Defines aggregating of multiple output scores.
    Array-like value defines weights used to average scor

In [33]:
from sklearn.linear_model import LinearRegression

regr_with_incerpet = LinearRegression(fit_intercept=True)
regr_with_incerpet.fit(diabetes_X, diabetes_y)

LinearRegression()

In [35]:
diabetes_y_pred_with_intercept = regr_with_incerpet.predict(diabetes_X)

In [36]:
# Coeficientes
print(f"Coefficients: \n{regr_with_incerpet.coef_}\n")
# Intercepto
print(f"Intercept: \n{regr_with_incerpet.intercept_}\n")
# Error cuadrático medio
print(f"Mean squared error: {mean_squared_error(diabetes_y, diabetes_y_pred_with_intercept):.2f}\n")
# Coeficiente de determinación
print(f"Coefficient of determination: {r2_score(diabetes_y, diabetes_y_pred_with_intercept):.2f}")

Coefficients: 
[ -10.01219782 -239.81908937  519.83978679  324.39042769 -792.18416163
  476.74583782  101.04457032  177.06417623  751.27932109   67.62538639]

Intercept: 
152.1334841628965

Mean squared error: 2859.69

Coefficient of determination: 0.52


In [37]:
regr_without_incerpet = LinearRegression(fit_intercept=False)
regr_without_incerpet.fit(diabetes_X, diabetes_y)

LinearRegression(fit_intercept=False)

In [38]:
diabetes_y_pred_without_intercept = regr_without_incerpet.predict(diabetes_X)

In [39]:
# Coeficientes
print(f"Coefficients: \n{regr_without_incerpet.coef_}\n")
# Error cuadrático medio
print(f"Mean squared error: {mean_squared_error(diabetes_y, diabetes_y_pred_without_intercept):.2f}\n")
# Coeficiente de determinación
print(f"Coefficient of determination: {r2_score(diabetes_y, diabetes_y_pred_without_intercept):.2f}")

Coefficients: 
[ -10.01219782 -239.81908937  519.83978679  324.39042769 -792.18416163
  476.74583782  101.04457032  177.06417623  751.27932109   67.62538639]

Mean squared error: 26004.29

Coefficient of determination: -3.39


**Pregunta: ¿Qué tan bueno fue el ajuste del modelo?**

__Respuesta:__
No es bueno en ninguno de los casos. Se ve un error cuadrático muy alto en el segundo caso, pero en ambos malos coeficientes de determinación también. 

## Pregunta 3

(1 pto)

Realizar multiples regresiones lineales utilizando una sola _feature_ a la vez. 

En cada iteración:

- Crea un arreglo `X`con solo una feature filtrando `X`.
- Crea un modelo de regresión lineal con intercepto.
- Ajusta el modelo anterior.
- Genera una predicción con el modelo.
- Calcula e imprime las métricas de la pregunta anterior.

In [ ]:
for col in # FIX ME PLEASE #:
    X_i = # FIX ME PLEASE #
    regr_i = # FIX ME PLEASE #
    regr_i.# FIX ME PLEASE #
    diabetes_y_pred_i = # FIX ME PLEASE #
    print(f"Feature: {col}")
    print(f"\tCoefficients: {# FIX ME PLEASE #}")
    print(f"\tIntercept: {# FIX ME PLEASE #")
    print(f"\tMean squared error: {# FIX ME PLEASE #):.2f}")
    print(f"\tCoefficient of determination: {# FIX ME PLEASE #i):.2f}\n")

**Pregunta: Si tuvieras que escoger una sola _feauture_, ¿Cuál sería? ¿Por qué?**

**Respuesta:** FIX ME

## Ejercicio 4

(1 pto)

Con la feature escogida en el ejercicio 3 realiza el siguiente gráfico:

- Scatter Plot
- Eje X: Valores de la feature escogida.
- Eje Y: Valores de la columna a predecir (target).
- En color rojo dibuja la recta correspondiente a la regresión lineal (utilizando `intercept_`y `coefs_`).
- Coloca un título adecuado, nombre de los ejes, etc.

Puedes utilizar `matplotlib` o `altair`, el que prefieras.

In [ ]:
regr = linear_model.LinearRegression(## FIX ME ##).fit(## FIX ME ##, ## FIX ME ##)

In [ ]:
## FIX ME ##